src/com/amazon/ion/impl/IonReaderBinaryRawX.java

```diff
@@ -848,15 +858,19 @@ private final long getPosition() {
         long pos = _input.getPosition();
         return pos;
     }
-    private final void skip(int len) throws IOException
+    private final void skip(long len) throws IOException
     {
         if (len < 0) {
             // no need to test this start >= dst.length ||
             // since we test start+len > dst.length which is the correct test
             throw new IllegalArgumentException();
         }
         if (_local_remaining == NO_LIMIT) {
-            _input.skip(len);
+            while (len > 0) {
+                int toSkip = (int) Math.min(Integer.MAX_VALUE, len);
+                _input.skip(toSkip);
+                len -= toSkip;
```

Comment:
[`InputStream#skip`](https://docs.oracle.com/javase/7/docs/api/java/io/InputStream.html#skip(long)) reserves the right to skip fewer bytes than were requested:
> Skips over and discards n bytes of data from this input stream. The skip method may, for a variety of reasons, end up skipping over some smaller number of bytes, possibly 0. This may result from any of a number of conditions; reaching end of file before n bytes have been skipped is only one possibility. The actual number of bytes skipped is returned. If n is negative, no bytes are skipped. 
```suggestion
                int toSkip = (int) Math.min(Integer.MAX_VALUE, len);
                int numberOfBytesSkipped = _input.skip(toSkip);
                len -= numberOfBytesSkipped;
```
What's not clear to me is whether it's possible for `skip` to return `0` forever, which would obviously be a problem inside `while (len > 0)`.

Reply:
In this case `_input` is a [`UnifiedInputStreamX`](https://github.com/amzn/ion-java/blob/master/src/com/amazon/ion/impl/UnifiedInputStreamX.java), which is not an `InputStream`. Its [`skip`](https://github.com/amzn/ion-java/blob/master/src/com/amazon/ion/impl/UnifiedInputStreamX.java#L330) method returns `void`.

There is no documentation on the method, but by looking at it, it's clear that it is up to the caller to make sure that the number of bytes requested to be skipped can actually be skipped. Otherwise it throws an exception for unexpected EOF, which is the correct behavior for this reader implementation.

Reply:
> In this case _input is a [UnifiedInputStreamX](https://github.com/amzn/ion-java/blob/master/src/com/amazon/ion/impl/UnifiedInputStreamX.java), which is not an InputStream. Its [skip](https://github.com/amzn/ion-java/blob/master/src/com/amazon/ion/impl/UnifiedInputStreamX.java#L330) method returns void.

That's what I was missing, thanks for pointing it out. 👍